### PR TITLE
Implement Bilateral Guided Gaussian Splatting.

### PR DESCRIPTION
**Pull request by jefequien**
Bilateral Guided Radiance Fields showed that bilateral grids are very effective at modeling exposure changes and other ISP effects. This PR ports their bilateral grid library onto Gaussian splatting. It seems quite effective at removing floaters in casually captured videos and improves color-corrected PSNR on the bilarf dataset.

https://github.com/user-attachments/assets/2f97afc4-c136-4e6f-9fe4-d8f1a7bc4ed0

https://github.com/user-attachments/assets/2d7942ea-49dc-41f8-ad80-0a11a2b30645

Original authors also expected this to work quite well. https://github.com/yuehaowang/bilarf/issues/1

TODO:
- [x] Match original paper's learning rate schedule exactly.

Does this belong in gsplat or is this more appropriate for nerfstudio?

### CODE REVIEW

**Comment by ichsan2895** [comment]
how to test it with gsplat-mcmc? Is it like this?
```
pip install git+https://github.com/jefequien/gsplat-mcmc@c4793ad154356d87065fa154aadbf37cf44e00db
python3 gsplat/examples/simple_trainer_mcmc.py
````

**Comment by jefequien** [comment]
Use `simple_trainer_mcmc.py` with the `--exp_opt` flag. 
```
    python simple_trainer_mcmc.py --eval_steps $EVAL_STEPS --save_steps $SAVE_STEPS --disable_viewer --data_factor $DATA_FACTOR \
        --init_type sfm \
        --cap_max $CAP_MAX \
        --max_steps $MAX_STEPS \
        --data_dir $SCENE_FOLDER/$SCENE/ \
        --render_traj_path $RENDER_TRAJ_PATH \
        --exp_opt \
        --result_dir results/3dgs_exposure_opt/$SCENE/
```

**Comment by ichsan2895** [comment]
Great, I will test later. I will inform the result in `renders` channel in `nerfstudio` _discord_ servers.

**Comment by ichsan2895** [comment]
I got error when run `simple_trainer_mcmc.py`
```
Traceback (most recent call last):
  File "/workspace/GSPLAT_EXPOPT/gsplat-mcmc/examples/simple_trainer_mcmc.py", line 30, in <module>
    from gsplat.utils.lib_bilagrid import BilateralGrid, slice
ModuleNotFoundError: No module named 'gsplat.utils'
```

**Comment by jefequien** [comment]
Oops. Missed an `__init__.py` file. Added.

**Comment by ichsan2895** [comment]
Whats going on to this PR? It really nice feature implementation for gsplat :)

**Comment by jefequien** [comment]
@ichsan2895 The nerfstudio team is working on adding it to splatfacto and @liruilong940607 will probably decide what to do with this PR when he gets back from vacation.
https://github.com/nerfstudio-project/nerfstudio/pull/3316

**Comment by jefequien** [comment]
Bilateral grids can be used to improve both nerfacto and splatfacto. Thus it makes sense for the method to live nerfstudio and share code between the two. This method exists entirely outside the core gsplat rasterization library but could live in gsplat for showcase/experimentation purposes.

**Comment by SharkWipf** [comment]
Ah, okay, makes sense, I didn't realize it applied to NeRFs too.
It's a bit odd how NeRFs are integrated and splats aren't, I guess it makes sense with how splats came later but it feels sub-optimal.  
I guess in theory the most sensible design would be to split nerfs off as well, and having a gsplat library/project, a nerf library/project, a common shared library for things like this, and Nerfstudio to tie it all together as just a way to mix and match different methods/techniques. But that'd be a major refactor and probably would break all external methods.

**Comment by Tao-11-chen** [comment]
@jefequien In my experiment, if your image is large, it will take huge video memory because the grid to be sampled at once will be torch.Size([num_pixels, 12, 8, 16, 16]), do you have any ideas on it?

**Comment by Tao-11-chen** [comment]
@jefequien Thanks very much for your great implementation. In my experiments, the grid sampling of large images will take unacceptable video memory because it will sample from a very large 5D grid with size (num_pixels_in_the_img, 12, 8, 16, 16), do you have any ideas on it? thanks in advance.

**Comment by Tao-11-chen** [comment]
> @jefequien Thanks very much for your great implementation. In my experiments, the grid sampling of large images will take unacceptable video memory because it will sample from a very large 5D grid with size (num_pixels_in_the_img, 12, 8, 16, 16), do you have any ideas on it? thanks in advance.

The problem is solved with the custom grid_sample function, which can largely reduce memory consumption in the case of sampling from the same bilateral_grid(image).

File: gsplat/util/lib_bilagrid.py
Comment (liruilong940607) [inline gsplat/util/lib_bilagrid.py:1]: I think we could put this file under `examples/`, as this is not something falls into the umbrella of gsplat

File: gsplat/util/__init__.py
Comment (liruilong940607) [inline gsplat/util/__init__.py:1]: Remove this util/ folder if we move the `lib_bilagrid` out

**Review by liruilong940607 — CHANGES_REQUESTED** [review]
LGTM overall! Just a minor request on moving the `lib_bilagrid.py` to the `examples/` folder

File: gsplat/util/__init__.py
Comment (jefequien) [reply gsplat/util/__init__.py:1]: Done

**Review by liruilong940607 — APPROVED** [review]
LGTM! thanks Jeffrey for bring the bilateral grid into gsplat!

**Pull request merged**